File: pythonProject/npg_monte_carlo_tree_search/monte_carlo_tree_search.py
from monte_carlo_tree_search.node import Node
from monte_carlo_tree_search.util.argmax import argmax
from nullpoga_cui.player import Action


class MonteCarloTreeSearch:

    @classmethod
    def train(cls, root_node: Node, simulation: int) -> None:
        root_node.expand()
        for _ in range(simulation):
            root_node.evaluate()

    @classmethod
    def select_action(cls, root_node: Node) -> Action:
        legal_actions = root_node.state.legal_actions()
        # nagai action一覧をここで出力する
        print('legal_actions', legal_actions)
        visit_list = [child.n for child in root_node.children]
        # 選択するアクションは
        
        return legal_actions[argmax(visit_list)]

File: packages/core/src/models/mcts/monte-carlo-tree-search.ts
import { Node } from './node';
import { State } from '../state';
import { Action } from '../action';
import { argmax } from './utils';

export class MonteCarloTreeSearch {
    static train(rootNode: Node, simulation: number): void {
        rootNode.expand();
        for (let i = 0; i < simulation; i++) {
            rootNode.evaluate();
        }
    }

    static selectAction(rootNode: Node): Action {
        const legalActions = rootNode.state.legalActions();
        const visitList = rootNode.children?.map(child => child.n) || [];
        return legalActions[argmax(visitList)];
    }
}